ledger/complete/mtrie/node/node.go

```diff
@@ -122,12 +123,92 @@ func NewInterimNode(height int, lchild, rchild *Node) *Node {
 	return n
 }
 
+// deepCopy
+func (n *Node) deepCopy() *Node {
+	var h hash.Hash
+	copy(h[:], n.hashValue[:])
+
+	var p ledger.Path
+	copy(p[:], n.path[:])
+
+	var lChildCopy, rChildCopy *Node
+	if n.lChild != nil {
+		lChildCopy = n.lChild.deepCopy()
+	}
+	if n.rChild != nil {
+		rChildCopy = n.rChild.deepCopy()
+	}
+
+	return &Node{
+		lChild:    lChildCopy,
+		rChild:    rChildCopy,
+		height:    n.height,
+		path:      p,
+		payload:   n.payload.DeepCopy(),
+		maxDepth:  n.maxDepth,
+		regCount:  n.regCount,
+		hashValue: h,
+	}
+}
+
 // computeAndStoreHash computes the node's hash value and
 // stores the result in the nodes internal `hashValue` field
 func (n *Node) computeAndStoreHash() {
 	n.hashValue = n.computeHash()
 }
 
+// Pruned would look at the child of a node and if it can be pruned
+// it constructs a new subtrie and return that one instead of n
+// otherwise it would return the node itself if no change is required
+func (n *Node) Pruned() (*Node, bool) {
+	if n.IsLeaf() {
+		return n, false
+	}
+
+	// if leaf return it as is
+	// if non leaf bubble up
+	lChildEmpty := true
+	rChildEmpty := true
+	if n.lChild != nil {
+		lChildEmpty = n.lChild.IsADefaultNode()
+	}
+	if n.rChild != nil {
+		rChildEmpty = n.rChild.IsADefaultNode()
+	}
+	if rChildEmpty && lChildEmpty {
+		// is like a leaf
+		return NewLeaf(ledger.Path{}, ledger.EmptyPayload(), n.height), true
+	}
+
+	// if childNode is non empty
+	if !lChildEmpty && rChildEmpty && n.lChild.IsLeaf() {
+		// TODO merge these two methods
+		l := n.lChild.deepCopy()
+		l.hashValue = hash.HashInterNode(l.hashValue, ledger.GetDefaultHashForHeight(l.height))
+		l.height = l.height + 1
+
+		// bubble up all children and return as parent
+		return l, true
```

Comment:
I am a bit concerned about this approach, as I suspect noteworthy GC overhead:
* here, we create the compactified leaf at height `height` first
* then, in each recursive step up the tree, we check that we could compactify, create a new compactified leaf at `height + 1`

I think we can assume that the majority of our leafs have quite a long compactified final tail, probably around 128 or even more. so each time we unallocate a register, we create 128 objects to be garbage collected. 

I I have played around with the code. Changing this is quite complicated, yet possible: here a rough draft https://github.com/onflow/flow-go/pull/1560

Reply:
I'll leave this for future improvements and not address it in this PR.